enclave/core/sgx/exception.c

```diff
@@ -288,6 +288,14 @@ void oe_real_exception_dispatcher(oe_context_t* oe_context)
         oe_exception_record.context->rbp = td->last_ssa_rbp;
     }
 
+    /* Update the state to prevent nested exceptions prior to this point
+     * that could cause oe_context on the stack and td states overwritten.
+     * That is, any exception occurs prior to this point will cause the
+     * following oe_enter performs the early return flow (i.e., oe_enter will
+     * early return if state is not OE_TD_STATE_SECOND_LEVEL_EXCEPTION_HANDLING
+     * for the case of nested exceptions) */
+    td->state = OE_TD_STATE_SECOND_LEVEL_EXCEPTION_HANDLING;
```

Comment:
To prevent reordering of statements/instructions by the compiler, there needs to be a compiler barrier just before this statement.